source/MRPython/MREmbeddedPython.cpp

```diff
@@ -102,71 +163,91 @@ bool EmbeddedPython::setupArgv( int argc, char** argv )
 
-bool EmbeddedPython::runScript( const std::filesystem::path& path )
-{
-    if ( !instance_().available_ || !isPythonScript( path ) )
-        return false;
+            while ( true )
+            {
+                { // Wait for source code.
+                    State s{};
+                    cv_.wait( guard, [&]{
+                        s = state_.load();
+                        return s == State::starting || s == State::stopThread;
+                    } );
+                    if ( s == State::stopThread )
+                        break;
+                    state_ = State::running; // Nobody waits for this, so don't notify?
+                }
 
-    std::ifstream ifs( path );
-    std::ostringstream oss;
-    oss << ifs.rdbuf();
-    ifs.close();
-    std::string str = oss.str();
-    return runString( str );
-}
+                lastRunSuccessful_ = false;
 
-bool EmbeddedPython::isPythonScript( const std::filesystem::path& path )
-{
-    std::error_code ec;
-    if ( !std::filesystem::exists( path, ec ) )
-        return false;
-    if ( !std::filesystem::is_regular_file( path, ec ) )
-        return false;
+                static bool initOk = init_();
+                if ( !initOk )
+                {
+                    spdlog::error( "Failied to initialize Python." );
+                }
+                else
+                {
+                    try
+                    {
+                        auto libDir = SystemPath::getPythonModulesDirectory();
+                        auto libDirStr = utf8string( libDir );
+                        MR::replaceInplace( libDirStr, "\\", "\\\\" ); // double protect first for C++ second for Python
+                        // Create an empty dictionary that will function as a namespace.
+                        std::string redirectScript =
+                            "import sys\n"
+                            "import redirector\n"
+                            "sys.stdout = redirector.stdout()\n"
+                            "sys.stderr = redirector.stderr()\n"
+                            "sys.path.insert(1,\"" + libDirStr + "\")\n";
+                        pybind11::exec( redirectScript.c_str() );
 
-    auto ext = utf8string( path.extension() );
-    for ( auto& c : ext )
-        c = (char) tolower( c );
+                        // Execute code
+                        pybind11::exec( queuedSource_.c_str() );
 
-    if ( ext != ".py" )
-        return false;
-    
-    return true;
-}
+                        pybind11::exec( "sys.stdout.flush()\nsys.stderr.flush()" );
 
-EmbeddedPython::EmbeddedPython()
-{
-    available_ = !Py_IsInitialized();
+                        // We used to reset globals using `pybind11::finalize_interpreter()`, but that beaks some modules (including ours at one point,
+                        //   and apparently numpy too (https://stackoverflow.com/q/7676314/2752075), so it's not necessarily our bug).
+                        pybind11::globals().clear();
```

Comment:
should we clean it before `try` or at the beginning to be sure that previous run will not interfere (if it has exceptions)?

Reply:
Moved after the catch, to ensure memory usage doesn't linger after the run.